examples_tests/50.MeshPacking/main.cpp

```diff
@@ -1,190 +1,388 @@
-    scene::ICameraSceneNode* camera = smgr->addCameraSceneNodeFPS(0, 100.0f, 0.5f);
-
-    camera->setPosition(core::vector3df(-4, 0, 0));
-    camera->setTarget(core::vector3df(0, 0, 0));
-    camera->setNearValue(1.f);
-    camera->setFarValue(5000.0f);
-
-    smgr->setActiveCamera(camera);
-
-    uint64_t lastFPSTime = 0;
-    while (device->run() && receiver.keepOpen())
-    {
-        driver->bindGraphicsPipeline(gpuPipeline.get());
-
-        driver->beginScene(true, true, video::SColor(255, 0, 0, 255));
-
-        //! This animates (moves) the camera and sets the transforms
-        camera->OnAnimate(std::chrono::duration_cast<std::chrono::milliseconds>(device->getTimer()->getTime()).count());
-        camera->render();
-
-        driver->pushConstants(gpuPipeline->getLayout(), asset::ISpecializedShader::ESS_VERTEX, 0u, sizeof(core::matrix4SIMD), camera->getConcatenatedMatrix().pointer());
-        driver->drawIndexedIndirect(mdiCallParams.vtxBindings, mdiCallParams.mode, mdiCallParams.indexType, mdiCallParams.indexBuff.get(), mdiCallParams.indirectDrawBuff.get(), mdiCallParams.offset, mdiCallParams.maxCount, mdiCallParams.stride);
-
-        driver->endScene();
-    }
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#define _NBL_STATIC_LIB_
+#include <iostream>
+#include <cstdio>
+#include <nabla.h>
+
+//! I advise to check out this file, its a basic input handler
+#include "../common/QToQuitEventReceiver.h"
+#include "nbl/asset/CCPUMeshPackerV1.h"
+#include "nbl/asset/CCPUMeshPackerV2.h"
+
+using namespace nbl;
+using namespace core;
+using namespace asset;
+using namespace video;
+
+constexpr const char* SHADER_OVERRIDES[2] =
+{
+R"(
+#define _NBL_VERT_INPUTS_DEFINED_
+#define _NBL_VERT_MAIN_DEFINED_
+
+//pos
+layout(set = 0, binding = 0) uniform samplerBuffer MeshPackedData_R32G32B32_SFLOAT;
+
+//uv
+//layout(set = 0, binding = 1) uniform samplerBuffer MeshPackedData_R32G32_SFLOAT;
+
+//normal
+layout(set = 0, binding = 2) uniform samplerBuffer MeshPackedData_A2B10G10R10_SNORM_PACK32;
+
+layout(set = 0, binding = 3) readonly buffer VertexDataOffsetTable
+{
+    uint dataOffsetTable[];
+} vertexPosition;
+
+layout(set = 0, binding = 4) readonly buffer VertexNormalOffsetTable
+{
+    uint dataOffsetTable[];
+} vertexNormal;
+
+)",
+
+R"(
+void main()
+{
+    vec3 pos = texelFetch(MeshPackedData_R32G32B32_SFLOAT,int(gl_VertexIndex)+vertexPosition.dataOffsetTable[gl_DrawID]).xyz;
+    LocalPos = pos;
+    gl_Position = nbl_glsl_pseudoMul4x4with3x1(CamData.params.MVP, pos);
+    ViewPos = nbl_glsl_pseudoMul3x4with3x1(CamData.params.MV, pos);
+    mat3 normalMat = nbl_glsl_SBasicViewParameters_GetNormalMat(CamData.params.NormalMatAndEyePos);
+
+    vec3 normal = texelFetch(MeshPackedData_A2B10G10R10_SNORM_PACK32,int(gl_VertexIndex)+vertexNormal.dataOffsetTable[gl_DrawID]).xyz;
```

Comment:
you'll need to have texture buffer format set as R32_UINT and decode rgb10a2 manually (theres already decode/encode for unorm rgb10a2 in glsl builtins (nbl/builtin/glsl/format/)), remember to normalize after decode

Reply:
https://www.khronos.org/registry/vulkan/specs/1.2/html/chap34.html#formats-mandatory-features-32bit

Table 47
`VK_FORMAT_A2B10G10R10_UNORM_PACK32` is mandatory

Only UNORM support is mandatory, SNORM is not
It could be decoded by just doing `decValue*2.0-vec3(1.0)`

Reply:
well, fine, but our loaders use snorm for normals